typing/scalar.mli

```diff
@@ -0,0 +1,359 @@
+(*   Copyright 2025 Jane Street Group LLC                                 *)
+(*                                                                        *)
+(*   All rights reserved.  This file is distributed under the terms of    *)
+(*   the GNU Lesser General Public License version 2.1, with the          *)
+(*   special exception on linking described in the file LICENSE.          *)
+(*                                                                        *)
+(**************************************************************************)
+
+(** This module defines the scalar types intrinsic to the OCaml compiler, and the
+    primitive operations defined on them.
+
+    A [Scalar.t] represents a particular OCaml type that represents a scalar value. It
+    might be tagged, boxed, or neither. There is also a type parameter for the locality of
+    the scalar value, which represents the location in which that boxed values are
+    allocated.
+
+    The important consideration is for a [Scalar.t] to represent all of the primitives
+    that we want to expose. The submodules are organized to make it easy for use different
+    subsets of scalars in different places. Some examples:
+
+    - Primitive arguments don't depend on the locality of their arguments, but the results
+    do.
+    - Some primitives only take integers, some take only floats, and Three_way_compare
+    takes any scalar type.
+    - The bytecode compiler wants to easily map unboxed/untagged values to their [value]
+    equivalents
+    - The middle-end wants to easily cast between any integraal values using only certain
+    primitives.
+*)
+
+type any_locality_mode = Any_locality_mode
+
+module Integer_comparison : sig
+  type t =
+    | Ceq
+    | Cne
+    | Clt
+    | Cgt
+    | Cle
+    | Cge
+
+  val to_string : t -> string
+
+  val swap : t -> t
+
+  val negate : t -> t
+end
+
+module Float_comparison : sig
+  type t =
+    | CFeq
+    | CFneq
+    | CFlt
+    | CFnlt
+    | CFgt
+    | CFngt
+    | CFle
+    | CFnle
+    | CFge
+    | CFnge
+
+  val to_string : t -> string
+
+  val swap : t -> t
+
+  val negate : t -> t
+end
+
+module Maybe_naked : sig
+  type ('a, 'b) t =
+    | Value of 'a
+    | Naked of 'b
+end
+
+module type S := sig
+  type 'a width
+
+  type nonrec 'a t = ('a width, any_locality_mode width) Maybe_naked.t
+
+  val all : any_locality_mode t list
+
+  val map : 'a t -> f:('a -> 'b) -> 'b t
+
+  val ignore_locality : 'a t -> any_locality_mode t
+
+  val width : _ t -> any_locality_mode width
+
+  val to_string : any_locality_mode t -> string
+
+  val sort : any_locality_mode t -> Jkind_types.Sort.Const.t
+end
+
+module type Integral_width_constants := sig
+  type 'a t
+
+  val int8 : _ t
+
+  val int16 : _ t
+
+  val int32 : any_locality_mode t
+
+  val int64 : any_locality_mode t
+
+  val int : _ t
+
+  val nativeint : any_locality_mode t
+end
+
+module type Integral_constants := sig
```

Comment:
I don't understand the intent of these module names.

1) It looks like the difference is that one has just the boxed things and the other has boxed and unboxed things, but I'm not sure what that has to do with "width".

2) The front-end generally uses the word "constant" to mean "literal" (see, e.g., how it is used in parsetree and typedtree).  I don't think you're using it in that sense here, and am not sure in what sense these are constants.  I think this module type is meant to represent the built-in types, so maybe something like "Integral_builtin_types" would be a good name? (unsure)

Probably comments about what they represent / how they are used would be good.